### PR TITLE
safety_limiter: allow escape motion from collision

### PR DESCRIPTION
**Pull request by at-wat**
merge after https://github.com/at-wat/neonavigation/pull/222

### CODE REVIEW

**Comment by at-wat** [comment]
## [#605] FAILED on indigo

<details><summary>Test failed</summary>


```
build/test_results/costmap_cspace/gtest-test_costmap_3d.xml: 6 tests
build/test_results/costmap_cspace/roslint-costmap_cspace.xml: 1 tests
build/test_results/joystick_interrupt/roslint-joystick_interrupt.xml: 1 tests
build/test_results/joystick_interrupt/rostest-test_test_joystick_interrupt_rostest.xml: 1 tests
build/test_results/joystick_interrupt/rosunit-test_joystick_interrupt.xml: 3 tests
build/test_results/map_organizer/roslint-map_organizer.xml: 1 tests
build/test_results/neonavigation_common/roslint-neonavigation_common.xml: 1 tests
build/test_results/obj_to_pointcloud/roslint-obj_to_pointcloud.xml: 1 tests
build/test_results/planner_cspace/gtest-test_blockmem_gridmap.xml: 3 tests
build/test_results/planner_cspace/gtest-test_cyclic_vec.xml: 7 tests
build/test_results/planner_cspace/gtest-test_grid_metric_converter.xml: 2 tests
build/test_results/planner_cspace/gtest-test_motion_cache.xml: 1 tests
build/test_results/planner_cspace/roslint-planner_cspace.xml: 1 tests
build/test_results/planner_cspace/rostest-test_test_abort_rostest.xml: 1 tests
build/test_results/planner_cspace/rostest-test_test_navigation_compat_rostest.xml: 1 tests
build/test_results/planner_cspace/rostest-test_test_navigation_rostest.xml: 1 tests
build/test_results/planner_cspace/rostest-test_test_preempt_rostest.xml: 1 tests
build/test_results/planner_cspace/rosunit-test_abort.xml: 1 tests
build/test_results/planner_cspace/rosunit-test_navigate.xml: 1 tests
build/test_results/planner_cspace/rosunit-test_preempt.xml: 1 tests
build/test_results/safety_limiter/roslint-safety_limiter.xml: 1 tests
build/test_results/safety_limiter/rostest-test_test_safety_limiter_compat_rostest.xml: 1 tests
build/test_results/safety_limiter/rostest-test_test_safety_limiter_rostest.xml: 1 tests
build/test_results/safety_limiter/rosunit-test_safety_limiter.xml: 6 tests, 0 errors, 1 failures, 0 skipped
build/test_results/track_odometry/gtest-test_tf_projection.xml: 1 tests
build/test_results/track_odometry/roslint-track_odometry.xml: 1 tests
build/test_results/track_odometry/rostest-test_test_track_odometry_rostest.xml: 1 tests
build/test_results/track_odometry/rosunit-test_track_odometry.xml: 1 tests
build/test_results/trajectory_tracker/gtest-test_trajectory_tracker_filter.xml: 3 tests
build/test_results/trajectory_tracker/roslint-trajectory_tracker.xml: 1 tests
build/test_results/trajectory_tracker/rostest-test_test_trajectory_recorder_rostest.xml: 1 tests
build/test_results/trajectory_tracker/rosunit-test_trajectory_recorder.xml: 1 tests
Summary: 55 tests, 0 errors, 1 failures, 0 skipped
```
</details>

**Comment by at-wat** [comment]
## [#606] PASSED on indigo

<details><summary>All tests passed</summary>


```
build/test_results/costmap_cspace/gtest-test_costmap_3d.xml: 6 tests
build/test_results/costmap_cspace/roslint-costmap_cspace.xml: 1 tests
build/test_results/joystick_interrupt/roslint-joystick_interrupt.xml: 1 tests
build/test_results/joystick_interrupt/rostest-test_test_joystick_interrupt_rostest.xml: 1 tests
build/test_results/joystick_interrupt/rosunit-test_joystick_interrupt.xml: 3 tests
build/test_results/map_organizer/roslint-map_organizer.xml: 1 tests
build/test_results/neonavigation_common/roslint-neonavigation_common.xml: 1 tests
build/test_results/obj_to_pointcloud/roslint-obj_to_pointcloud.xml: 1 tests
build/test_results/planner_cspace/gtest-test_blockmem_gridmap.xml: 3 tests
build/test_results/planner_cspace/gtest-test_cyclic_vec.xml: 7 tests
build/test_results/planner_cspace/gtest-test_grid_metric_converter.xml: 2 tests
build/test_results/planner_cspace/gtest-test_motion_cache.xml: 1 tests
build/test_results/planner_cspace/roslint-planner_cspace.xml: 1 tests
build/test_results/planner_cspace/rostest-test_test_abort_rostest.xml: 1 tests
build/test_results/planner_cspace/rostest-test_test_navigation_compat_rostest.xml: 1 tests
build/test_results/planner_cspace/rostest-test_test_navigation_rostest.xml: 1 tests
build/test_results/planner_cspace/rostest-test_test_preempt_rostest.xml: 1 tests
build/test_results/planner_cspace/rosunit-test_abort.xml: 1 tests
build/test_results/planner_cspace/rosunit-test_navigate.xml: 1 tests
build/test_results/planner_cspace/rosunit-test_preempt.xml: 1 tests
build/test_results/safety_limiter/roslint-safety_limiter.xml: 1 tests
build/test_results/safety_limiter/rostest-test_test_safety_limiter_compat_rostest.xml: 1 tests
build/test_results/safety_limiter/rostest-test_test_safety_limiter_rostest.xml: 1 tests
build/test_results/safety_limiter/rosunit-test_safety_limiter.xml: 7 tests
build/test_results/track_odometry/gtest-test_tf_projection.xml: 1 tests
build/test_results/track_odometry/roslint-track_odometry.xml: 1 tests
build/test_results/track_odometry/rostest-test_test_track_odometry_rostest.xml: 1 tests
build/test_results/track_odometry/rosunit-test_track_odometry.xml: 1 tests
build/test_results/trajectory_tracker/gtest-test_trajectory_tracker_filter.xml: 3 tests
build/test_results/trajectory_tracker/roslint-trajectory_tracker.xml: 1 tests
build/test_results/trajectory_tracker/rostest-test_test_trajectory_recorder_rostest.xml: 1 tests
build/test_results/trajectory_tracker/rosunit-test_trajectory_recorder.xml: 1 tests
Summary: 56 tests, 0 errors, 0 failures, 0 skipped
```
</details>

File: safety_limiter/src/safety_limiter.cpp [hunk 1]
```diff
@@ -377,17 +392,35 @@ class SafetyLimiterNode
...
+        {
+          // The robot is already in collision.
+          // Allow movement under d_escape_ and yaw_escape_
+          d_escape_remain = d_escape_;
+          yaw_escape_remain = yaw_escape_;
+          has_collision_at_now = true;
+        }
+        if (d_escape_remain <= 0 || yaw_escape_remain <= 0)
+        {
+          if (has_collision_at_now)
```
Comment (DaikiMaekawa) [inline safety_limiter/src/safety_limiter.cpp]: preferably, I'd like to add {} here.

File: safety_limiter/src/safety_limiter.cpp [hunk 2]
```diff
@@ -344,13 +348,23 @@ class SafetyLimiterNode
...
+    float d_escape_remain = 0;
+    float yaw_escape_remain = 0;
+
     for (float t = 0; t < tmax_; t += dt_)
     {
-      d_col += twist_.linear.x * dt_;
-      yaw_col += twist_.angular.z * dt_;
-      move = move * motion;
-      move_inv = move_inv * motion_inv;
+      if (t != 0)
```
[the remark this answers is not in the document]
Comment (at-wat) [reply safety_limiter/src/safety_limiter.cpp]: separated PR is nicer

File: safety_limiter/src/safety_limiter.cpp [hunk 3]
```diff
@@ -377,17 +392,35 @@ class SafetyLimiterNode
...
           pos.z = p.z;
           col_points.points.push_back(pos);
-          col = true;
+          colliding = true;
+          break;
+        }
+      }
+      if (colliding)
+      {
+        if (t == 0)
```
[the remark this answers is not in the document]
Comment (at-wat) [reply safety_limiter/src/safety_limiter.cpp]: this is not a part of the above bug, but an impl. of collision escaping logic.

File: safety_limiter/src/safety_limiter.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (at-wat) [reply safety_limiter/src/safety_limiter.cpp]: https://github.com/at-wat/neonavigation/pull/222

File: safety_limiter/src/safety_limiter.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (DaikiMaekawa) [reply safety_limiter/src/safety_limiter.cpp]: i got it. thx.

**Comment by at-wat** [comment]
## [#608] PASSED on indigo

<details><summary>All tests passed</summary>


```
build/test_results/costmap_cspace/gtest-test_costmap_3d.xml: 6 tests
build/test_results/costmap_cspace/roslint-costmap_cspace.xml: 1 tests
build/test_results/joystick_interrupt/roslint-joystick_interrupt.xml: 1 tests
build/test_results/joystick_interrupt/rostest-test_test_joystick_interrupt_rostest.xml: 1 tests
build/test_results/joystick_interrupt/rosunit-test_joystick_interrupt.xml: 3 tests
build/test_results/map_organizer/roslint-map_organizer.xml: 1 tests
build/test_results/neonavigation_common/roslint-neonavigation_common.xml: 1 tests
build/test_results/obj_to_pointcloud/roslint-obj_to_pointcloud.xml: 1 tests
build/test_results/planner_cspace/gtest-test_blockmem_gridmap.xml: 3 tests
build/test_results/planner_cspace/gtest-test_cyclic_vec.xml: 7 tests
build/test_results/planner_cspace/gtest-test_grid_metric_converter.xml: 2 tests
build/test_results/planner_cspace/gtest-test_motion_cache.xml: 1 tests
build/test_results/planner_cspace/roslint-planner_cspace.xml: 1 tests
build/test_results/planner_cspace/rostest-test_test_abort_rostest.xml: 1 tests
build/test_results/planner_cspace/rostest-test_test_navigation_compat_rostest.xml: 1 tests
build/test_results/planner_cspace/rostest-test_test_navigation_rostest.xml: 1 tests
build/test_results/planner_cspace/rostest-test_test_preempt_rostest.xml: 1 tests
build/test_results/planner_cspace/rosunit-test_abort.xml: 1 tests
build/test_results/planner_cspace/rosunit-test_navigate.xml: 1 tests
build/test_results/planner_cspace/rosunit-test_preempt.xml: 1 tests
build/test_results/safety_limiter/roslint-safety_limiter.xml: 1 tests
build/test_results/safety_limiter/rostest-test_test_safety_limiter_compat_rostest.xml: 1 tests
build/test_results/safety_limiter/rostest-test_test_safety_limiter_rostest.xml: 1 tests
build/test_results/safety_limiter/rosunit-test_safety_limiter.xml: 7 tests
build/test_results/track_odometry/gtest-test_tf_projection.xml: 1 tests
build/test_results/track_odometry/roslint-track_odometry.xml: 1 tests
build/test_results/track_odometry/rostest-test_test_track_odometry_rostest.xml: 1 tests
build/test_results/track_odometry/rosunit-test_track_odometry.xml: 1 tests
build/test_results/trajectory_tracker/gtest-test_trajectory_tracker_filter.xml: 3 tests
build/test_results/trajectory_tracker/roslint-trajectory_tracker.xml: 1 tests
build/test_results/trajectory_tracker/rostest-test_test_trajectory_recorder_rostest.xml: 1 tests
build/test_results/trajectory_tracker/rosunit-test_trajectory_recorder.xml: 1 tests
Summary: 56 tests, 0 errors, 0 failures, 0 skipped
```
</details>

**Comment by at-wat** [comment]
## [#610] PASSED on indigo

<details><summary>All tests passed</summary>


```
build/test_results/costmap_cspace/gtest-test_costmap_3d.xml: 6 tests
build/test_results/costmap_cspace/roslint-costmap_cspace.xml: 1 tests
build/test_results/joystick_interrupt/roslint-joystick_interrupt.xml: 1 tests
build/test_results/joystick_interrupt/rostest-test_test_joystick_interrupt_rostest.xml: 1 tests
build/test_results/joystick_interrupt/rosunit-test_joystick_interrupt.xml: 3 tests
build/test_results/map_organizer/roslint-map_organizer.xml: 1 tests
build/test_results/neonavigation_common/roslint-neonavigation_common.xml: 1 tests
build/test_results/obj_to_pointcloud/roslint-obj_to_pointcloud.xml: 1 tests
build/test_results/planner_cspace/gtest-test_blockmem_gridmap.xml: 3 tests
build/test_results/planner_cspace/gtest-test_cyclic_vec.xml: 7 tests
build/test_results/planner_cspace/gtest-test_grid_metric_converter.xml: 2 tests
build/test_results/planner_cspace/gtest-test_motion_cache.xml: 1 tests
build/test_results/planner_cspace/roslint-planner_cspace.xml: 1 tests
build/test_results/planner_cspace/rostest-test_test_abort_rostest.xml: 1 tests
build/test_results/planner_cspace/rostest-test_test_navigation_compat_rostest.xml: 1 tests
build/test_results/planner_cspace/rostest-test_test_navigation_rostest.xml: 1 tests
build/test_results/planner_cspace/rostest-test_test_preempt_rostest.xml: 1 tests
build/test_results/planner_cspace/rosunit-test_abort.xml: 1 tests
build/test_results/planner_cspace/rosunit-test_navigate.xml: 1 tests
build/test_results/planner_cspace/rosunit-test_preempt.xml: 1 tests
build/test_results/safety_limiter/roslint-safety_limiter.xml: 1 tests
build/test_results/safety_limiter/rostest-test_test_safety_limiter_compat_rostest.xml: 1 tests
build/test_results/safety_limiter/rostest-test_test_safety_limiter_rostest.xml: 1 tests
build/test_results/safety_limiter/rosunit-test_safety_limiter.xml: 7 tests
build/test_results/track_odometry/gtest-test_tf_projection.xml: 1 tests
build/test_results/track_odometry/roslint-track_odometry.xml: 1 tests
build/test_results/track_odometry/rostest-test_test_track_odometry_rostest.xml: 1 tests
build/test_results/track_odometry/rosunit-test_track_odometry.xml: 1 tests
build/test_results/trajectory_tracker/gtest-test_trajectory_tracker_filter.xml: 3 tests
build/test_results/trajectory_tracker/roslint-trajectory_tracker.xml: 1 tests
build/test_results/trajectory_tracker/rostest-test_test_trajectory_recorder_rostest.xml: 1 tests
build/test_results/trajectory_tracker/rosunit-test_trajectory_recorder.xml: 1 tests
Summary: 56 tests, 0 errors, 0 failures, 0 skipped
```
</details>

**Pull request merged**